openmc/mgxs/library.py

```diff
@@ -1005,32 +1018,79 @@ def create_mg_library(self, xs_type='macro', xsdata_names=None,
         # Initialize file
         mgxs_file = openmc.MGXSLibrary(self.energy_groups)
 
-        # Create the xsdata object and add it to the mgxs_file
-        for i, domain in enumerate(self.domains):
-            if self.by_nuclide:
-                nuclides = list(domain.get_all_nuclides().keys())
+        if xs_ids is not None:
+            if isinstance(xs_ids, basestring):
+                # If we only have a string lets convert it now to a list
+                # of strings.
+                xs_id_input = xs_ids
+                xs_ids = []
+                for domain in self.domains:
+                    if self.domain_type is 'mesh':
+                        for subdomain in domain.cell_generator():
+                            xs_ids.append(xs_id_input)
+                    else:
+                        xs_ids.append(xs_id_input)
             else:
-                nuclides = ['total']
-            for nuclide in nuclides:
-                # Build & add metadata to XSdata object
-                if xsdata_names is None:
-                    xsdata_name = 'set' + str(i + 1)
+                cv.check_iterable_type('xs_ids', xs_ids, basestring)
+        else:
+            xs_ids = []
+            for domain in self.domains:
+                if self.domain_type is 'mesh':
+                    for subdomain in domain.cell_generator():
+                        xs_ids.append('1m')
                 else:
-                    xsdata_name = xsdata_names[i]
-                if nuclide is not 'total':
-                    xsdata_name += '_' + nuclide
+                    xs_ids.append('1m')
 
-                xsdata = self.get_xsdata(domain, xsdata_name, nuclide=nuclide,
-                                         xs_type=xs_type, xs_id=xs_ids[i],
-                                         tabular_legendre=tabular_legendre,
-                                         tabular_points=tabular_points)
+        if self.domain_type == 'mesh':
+            # Create the xsdata objects and add to the mgxs_file
+            i = 0
+            for domain in self.domains:
+                for subdomain in domain.cell_generator():
```

Comment:
Ditto, I would recommend avoiding the mesh cell iteration unless explicitly needed.